### PR TITLE
Add CUDA support for the OFI MTL

### PR DESCRIPTION
**Pull request by wckzhang**
This patch series touches the CUDA, opal datatype, PML CM, MTL datatype, and OFI MTL code.

The important things added here are:
Public function to allocate CUDA memory for datapacking of heterogenous buffers and device-to-device copy.
Pointer attribute detection in the CM PML layer.
Selecting libfabric providers with FI_HMEM and FI_MR_HMEM capabilities.
Only compiling the OFI MTL if libfabric version >= 1.9 if CUDA support is requested.
Memory registration for CUDA buffers.

### CODE REVIEW

File: opal/datatype/opal_datatype_cuda.c [hunk 1]
```diff
@@ -111,6 +111,59 @@ bool  opal_cuda_check_one_buf(char *buf, opal_convertor_t *convertor )
     return ( ftable.gpu_is_gpu_buffer(buf, convertor));
 }
 
+/*
```
Comment (bosilca) [inline opal/datatype/opal_datatype_cuda.c:114]: The OPAL datatype packing and unpacking is too low level to provide support for memory allocation. I understand this seems to be a popular place to put CUDA functions, but I would suggest these functions move to opal/mca/common/cuda.

File: opal/datatype/opal_datatype_cuda.h [hunk 2]
```diff
@@ -18,12 +18,16 @@ struct opal_common_cuda_function_table {
     int (*gpu_cu_memcpy_async)(void*, const void*, size_t, opal_convertor_t*);
     int (*gpu_cu_memcpy)(void*, const void*, size_t);
     int (*gpu_memmove)(void*, void*, size_t);
+    int (*gpu_malloc)(void*, size_t);
+    int (*gpu_free)(void*);
```
Comment (bosilca) [inline opal/datatype/opal_datatype_cuda.h:22]: Technically I think it makes much more sense to move the entire CUDA structure outside of the datatype (especially now that it start having additional capabilities).

**Review by bosilca — APPROVED** [review]

File: opal/datatype/opal_datatype_cuda.h (hunk 2 above)
Comment (wckzhang) [reply opal/datatype/opal_datatype_cuda.h:22]: Hmm, thanks for the feedback, I'll see what I can do about it.

**Comment by gpaulsen** [comment]
@rajachan, you mentioned on Tuesday that AWS is doing some additional evaluation on this PR, is that correct?

**Review by hppritcha — APPROVED** [review]
I approve but agree with George's comments.

**Comment by wckzhang** [comment]
Updated, there was a minor bug in the malloc code which I fixed (Didn't use a double pointer), and just basically moved all the opal_datatype_cuda.c and opal_datatype_cuda.h files into the corresponding common_cuda files.

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/970849af86693490a4531136ac9e4be8

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/e13faff41eb79d6f2230ba9dfca596e4

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/8423dc40cc266d16a37d32334aa7c6b9

**Comment by wckzhang** [comment]
Looks like there's something I'm missing with compilation. Didn't hit this error during personal testing, not sure what's different in the CI setup compared to the manual build.

```
make[2]: *** No rule to make target 'mca/common/cuda/libmca_common_cuda.la', needed by 'libopen-pal.la'.  Stop.
```

**Comment by gpaulsen** [comment]
@wckzhang did you figure it out?  I was going to try a local ppc64le build if you're stuck.

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/26475e9bedc1d8ff0754ee6cbe4ac127

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/cc495858babc4b613fdf123abf6e89c3

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ba79b0db3c1bbe0726e4cd6e546ce0bf

**Comment by wckzhang** [comment]
I fixed the first error have another bug with cuda.h now it looks like

**Comment by wckzhang** [comment]
oh yeah I'm pretty sure I know what's up, I only tested with cuda compiled, will add checks only if compiled with cuda.

**Comment by wckzhang** [comment]
Bit of a dumb solution but it compiled properly both with and without cuda.

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/eea625079b1fa0922b8b1763d51206f0

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/3f258c11df63d27640631c009cc9bae8

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/e13e080e4383fc2410059a3d23bae826

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/04c803512823728b34dc32f84746452f

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/75d0a227795ea5b0d1fa4ba342c09735

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ff4dbb87c599b309a70414807dccd5c0

**Comment by wckzhang** [comment]
Hmm, I don't see what the error is this time in the output, are the files truncated?

**Comment by jjhursey** [comment]
It looks like it hit a 20-minute timeout on the make stage. Let me adjust that then I'll ask for a retest

**Comment by jjhursey** [comment]
bot:ibm:retest

**Review by bosilca — APPROVED** [review]

**Review by hppritcha — APPROVED** [review]

**Comment by hppritcha** [comment]
Unfortunately this PR breaks a bunch of IBM tests which involve derived, non-contiguous datatypes when using the OFI mtl.

File: ompi/mca/pml/cm/pml_cm_sendreq.h [hunk 3]
```diff
@@ -242,6 +242,14 @@ do {                                                                    \
             (unsigned char*)buf + datatype->super.true_lb;              \
         (req_send)->req_base.req_convertor.count      = count;          \
         (req_send)->req_base.req_convertor.pDesc      = &datatype->super; \
+        /* Switches off CUDA detection if                               \
+         MTL set MCA_MTL_BASE_FLAG_CUDA_INIT_DISABLE during init */     \
+        MCA_PML_CM_SWITCH_CUDA_CONVERTOR_OFF(flags, datatype, count);   \
+        (req_send)->req_base.req_convertor.flags |= flags;              \
+        /* Sets CONVERTOR_CUDA flag if CUDA buffer */                   \
+        opal_convertor_prepare_for_send(                                \
```
Comment (hppritcha) [inline ompi/mca/pml/cm/pml_cm_sendreq.h:250]: this broke use of OFI MTL in when not supporting CUDA.

Comment (wckzhang) [reply ompi/mca/pml/cm/pml_cm_sendreq.h:250]: Can you describe what the exact test matrix was, ie. OMPI compiled w/ CUDA using a provider that supports/doesn't support FI_HMEM, using host buffers?

Comment (wckzhang) [reply ompi/mca/pml/cm/pml_cm_sendreq.h:250]: Is this the same issue as your previous comment? This segment seems to only affect contiguous memory it looks like, does removing this section of code fix your non-contiguous use case?

Comment (bosilca) [reply ompi/mca/pml/cm/pml_cm_sendreq.h:250]: Please read my comment on #8906 about this code.

**Pull request merged**